API/Program.cs

```diff
@@ -71,6 +80,82 @@
+            RateLimitConfiguration.Position);
+
+        var defaultFixedOptions = new FixedWindowRateLimiterOptions()
+        {
+            PermitLimit = 30,
+            Window = TimeSpan.FromSeconds(60),
+            QueueLimit = 0
+        };
+
+        // Configure response for rate limited clients
+        options.OnRejected = async (context, token) =>
+        {
+            context.HttpContext.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+            const string readMore =
+                "Read more about our rate limits at https://github.com/osu-tournament-rating/otr-wiki/blob/master/api/limits/en.md";
+            if (context.Lease.TryGetMetadata(MetadataName.RetryAfter, out TimeSpan retryAfter))
+            {
+                await context.HttpContext.Response.WriteAsync(
+                    $"Too many requests. Please try again after {retryAfter.TotalSeconds} second(s). " + readMore,
+                    token
+                );
+            }
+            else
+            {
+                await context.HttpContext.Response.WriteAsync(
+                    "Too many requests. Please try again later. " + readMore, token);
+            }
+        };
+
+
+        options.GlobalLimiter = PartitionedRateLimiter.Create<HttpContext, string>(context =>
+        {
+            // Try to get access token
+            var accessToken =
+                context.Features.Get<IAuthenticateResultFeature>()?.AuthenticateResult?.Properties
+                    ?.GetTokenValue("access_token")?.ToString() ?? string.Empty;
+
+            // Shared partition for anonymous users
+            if (accessToken.IsNullOrEmpty())
+            {
+                return RateLimitPartition.GetFixedWindowLimiter(
+                    "anonymous",
+                    _ => defaultFixedOptions
+                );
+            }
+
+            // Try parse rate limit override claim
+            JwtSecurityToken jwtToken = new JwtSecurityTokenHandler().ReadJwtToken(accessToken);
+            var overrideClaim = jwtToken.Claims.FirstOrDefault(c => c.Type == "ratelimitoverrides")?.Value;
+            RateLimitOverrides? overrides = null;
+            if (!string.IsNullOrEmpty(overrideClaim))
+            {
+                overrides = JsonSerializer.Deserialize<RateLimitOverrides>(overrideClaim);
+            }
+
+            // Partition for each unique user / client
+            if (overrides is null)
+            {
+                return RateLimitPartition.GetFixedWindowLimiter(jwtToken.Issuer, _ => defaultFixedOptions);
+            }
+            return RateLimitPartition.GetFixedWindowLimiter(
+                jwtToken.Issuer,
+                _ => new FixedWindowRateLimiterOptions()
+                {
+                    PermitLimit = overrides.PermitLimit ?? defaultFixedOptions.PermitLimit,
+                    Window = TimeSpan.FromSeconds(overrides.Window ?? rateLimitConfiguration.Window),
```

Comment:
On one line `defaultFixedOptions` is used, the next line `rateLimitConfiguration` is used...?

Reply:
My thought is that it would make more sense to use the earlier defined values from `defaultFixedOptions` for consistency, but since `defaultFixedOptions.Window` is already parsed into a `TimeSpan` in order to use a nullish coalescence on `overrides.Window` I would need to use the config value instead. For readability I will just use the values from `rateLimitConfiguration` instead.